components/raftstore-v2/src/operation/command/admin/merge/mod.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+pub mod prepare;
+
+use engine_traits::{KvEngine, RaftEngine};
+use kvproto::{
+    raft_cmdpb::RaftCmdRequest,
+    raft_serverpb::{PeerState, RegionLocalState},
+};
+use prepare::PrepareStatus;
+use raft::{ProgressState, INVALID_INDEX};
+use raftstore::Result;
+use slog::{info, warn, Logger};
+use tikv_util::box_err;
+
+use crate::raft::Peer;
+
+#[derive(Default)]
+pub struct MergeContext {
+    prepare_status: Option<PrepareStatus>,
+}
+
+impl MergeContext {
+    #[inline]
+    pub fn from_region_state(logger: &Logger, state: &RegionLocalState) -> Option<Self> {
+        if state.get_state() == PeerState::Merging {
+            info!(logger, "region is merging"; "region_state" => ?state);
+            let mut ctx = Self::default();
+            ctx.prepare_status = Some(PrepareStatus::Applied(state.get_merge_state().clone()));
+            Some(ctx)
+        } else {
+            None
+        }
+    }
+
+    #[inline]
+    pub fn maybe_take_pending_prepare(&mut self, applied: u64) -> Option<RaftCmdRequest> {
+        if let Some(PrepareStatus::WaitForFence {
+            fence,
+            req,
+            ..
+        }) = self.prepare_status.as_mut()
+            && applied >= *fence
+        {
+            // The status will be updated during processing the proposal.
+            return req.take();
+        }
+        None
+    }
+
+    #[inline]
+    pub fn max_compact_log_index(&self) -> Option<u64> {
+        if let Some(PrepareStatus::WaitForFence { ctx, .. }) = self.prepare_status.as_ref() {
+            Some(ctx.min_matched)
+        } else {
+            None
+        }
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    #[inline]
+    pub fn update_merge_progress_on_became_follower(&mut self) {
+        self.take_merge_context();
```

Comment:
It should only clear `PrepareStatus::WaitForFence`. All other states should be preserved. And proposal_control should also be updated.